controllers/routing/routing_delete_resources.go

```diff
@@ -2,45 +2,70 @@ package routing
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
-// HandleResourceDeletion handles the removal of dependent resources when the target resource is being deleted.
-func (r *PlatformRoutingController) HandleResourceDeletion(ctx context.Context, sourceRes *unstructured.Unstructured) (ctrl.Result, error) {
-	exportModes, found := extractExportModes(sourceRes)
-	if !found {
+func (r *PlatformRoutingController) removeUnusedRoutingResources(ctx context.Context, target *unstructured.Unstructured) error {
+	if !target.GetDeletionTimestamp().IsZero() {
+		return r.handleResourceDeletion(ctx, target)
+	}
+
+	exportModes := extractExportModes(target)
+	unusedRouteTypes := spi.UnusedRouteTypes(exportModes)
+
+	var errDeletion []error
+
+	for _, unusedRouteType := range unusedRouteTypes {
+		if errDel := r.deleteOwnedResources(ctx, target, unusedRouteType, routingResourceGVKs(unusedRouteType)); errDel != nil {
+			errDeletion = append(errDeletion, errDel)
+		}
+	}
+
+	return errors.Join(errDeletion...)
+}
+
+// handleResourceDeletion handles the removal of dependent resources when the target resource is being deleted.
+func (r *PlatformRoutingController) handleResourceDeletion(ctx context.Context, sourceRes *unstructured.Unstructured) error {
+	exportModes := extractExportModes(sourceRes)
+	if len(exportModes) == 0 {
 		r.log.Info("No export modes found, skipping deletion logic", "sourceRes", sourceRes)
 
-		return ctrl.Result{}, nil
+		return nil
 	}
 
 	r.log.Info("Handling deletion of dependent resources", "sourceRes", sourceRes)
 
 	for _, exportMode := range exportModes {
-		if err := r.deleteOwnedResources(ctx, sourceRes, routingResourceGVKs(exportMode)); err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to delete resources for export mode %s: %w", exportMode, err)
+		if err := r.deleteOwnedResources(ctx, sourceRes, exportMode, routingResourceGVKs(exportMode)); err != nil {
+			return fmt.Errorf("failed to delete resources for export mode %s: %w", exportMode, err)
 		}
 	}
 
 	return removeFinalizer(ctx, r.Client, sourceRes)
 }
 
-func (r *PlatformRoutingController) deleteOwnedResources(ctx context.Context, target *unstructured.Unstructured, gvkList []schema.GroupVersionKind) error {
+func (r *PlatformRoutingController) deleteOwnedResources(ctx context.Context,
+	target *unstructured.Unstructured,
+	exportMode spi.RouteType,
+	gvkList []schema.GroupVersionKind) error {
 	ownerName := target.GetName()
 	ownerKind := target.GetObjectKind().GroupVersionKind().Kind
 	ownerUID := string(target.GetUID())
+	exportType := string(exportMode)
 
 	resourceOwnerLabels := client.MatchingLabels{
-		metadata.Labels.OwnerName: ownerName,
-		metadata.Labels.OwnerKind: ownerKind,
-		metadata.Labels.OwnerUID:  ownerUID,
+		metadata.Labels.OwnerName:  ownerName,
+		metadata.Labels.OwnerKind:  ownerKind,
+		metadata.Labels.OwnerUID:   ownerUID,
+		metadata.Labels.ExportType: exportType,
```

Comment:
Technically speaking this is not part of "resource owner" but see my comment above about the label selector.

Reply:
agree, moved as part of other review